boards: arc: Update EM StarterKit board support from 2.2 to 2.3

fanghuaqi opened the pull request:
EM Starterkit(namely EMSK) board has released out a new firmware version called 2.03a(namely the 2.3), 
The main changes of this new firmware list as follows:
+ Support for ARC EM 4.00a, including the following enhanced options compared to EMSK 2.2:
   - Enhanced Security Package (ESP) SecureShield™ extension for secure mode.
   - µDMA.
+ DMA support for DesignWare SPI master.
+ ARC EM7D MPU update to support secure mode and multiple context IDs.

We update software support in Zephyr to support this new version of EMSK.

The main updates of the software for EMSK 2.3 list as follows:
* board: Update EMSK onboard resources such as Button, Switch and LEDs
  + update soc.h for em7d, em9d, em11d
  + update board.h for em_starterkit board
* arc: Add floating point support and code density support
  + add kconfig configuration
  + add compiler options
  + add register definitions, marcos, assembly codes
  + fixes in existing codes and configurations.
* arc: Update detailed board configurations for cores of emsk 2.3
* arc: Only support em9d and em11d. 
  + em7d is a core with secure feature, which is not supported in Zephyr, if you want to access to try secureshield feature, you can go to [embARC OSP repo](https://github.com/foss-for-synopsys-dwc-arc-processors/embarc_osp), fork and try.
* script: Provide arc_debugger.sh for debugging em_starterkit board
  + make BOARD=em_starterkit debug
    This will start openocd server for emsk, and arc gdb will connect
    to this debug server, user can run `continue` command if user just
    want to run the application, or other commands if debugging needed.
  + make BOARD=em_starterkit debugserver
    This will start an openocd debugger server for emsk, and user can
    connect to this debugserver using arc gdb and do what they want to.
  + make BOARD=em_starterkit flash
    This will download the zephyr application elf file to emsk,
    and run it.

fanghuaqi committed:
boards: Update arc em_starterkit support from 2.2 to 2.3

Here are the main changes:
* board: Update EMSK onboard resources such as Button, Switch and LEDs
  + update soc.h for em7d, em9d, em11d
  + update board.h for em_starterkit board
* arc: Add floating point support and code density support
  + add kconfig configuration
  + add compiler options
  + add register definitions, marcos, assembly codes
  + fixes in existing codes and configurations.
* arc: Update detailed board configurations for cores of emsk 2.3
* script: Provide arc_debugger.sh for debugging em_starterkit board
  + make BOARD=em_starterkit debug
    This will start openocd server for emsk, and arc gdb will connect
    to this debug server, user can run `continue` command if user just
    want to run the application, or other commands if debugging needed.
  + make BOARD=em_starterkit debugserver
    This will start an openocd debugger server for emsk, and user can
    connect to this debugserver using arc gdb and do what they want to.
  + make BOARD=em_starterkit flash
    This will download the zephyr application elf file to emsk,
    and run it.

Signed-off-by: Huaqi Fang <huaqi.fang@synopsys.com>

Commit:
arc: Fix the bug in _load_callee_saved_regs

load floating register first, or the value of r13 is not correct

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

fanghuaqi committed:
boards: Make em9d to the default core configuration for emsk

Signed-off-by: Huaqi Fang <huaqi.fang@synopsys.com>

fanghuaqi committed:
doc: Update emsk board debug and flash section

Due to support updated from 2.2 to 2.3, a debug script is provided
for emsk, so developer can directly debug the zephyr application
using commands such as follows.
+ make BOARD=em_starterkit debug

Signed-off-by: Huaqi Fang <huaqi.fang@synopsys.com>

tbursztyka commented on arch/arc/soc/em11d/soc.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016 Synopsys, Inc. All rights reserved.
+ * Copyright (c) 2017 Synopsys, Inc. All rights reserved.
```
2016-2017

pfalcon replied:
> Unless I am mistaken, if you want to update the date keep the very first one and add the new one besides: s/2016/2016-2017

I believe per the copyright law, last date of work is important: 50 years after that, it's in public domain. Starting date is just a trivia for curious (which I personally appreciate).

fanghuaqi replied:
Thanks for kindly notice, I will modify it.

tbursztyka commented on boards/arc/em_starterkit/Makefile.board:
```diff
@@ -1,16 +1,7 @@
-OPENOCD_LOAD_CMD = "load_image ${O}/${KERNEL_BIN_NAME} $(CONFIG_ICCM_BASE_ADDRESS)"
-OPENOCD_VERIFY_CMD = "verify_image ${O}/${KERNEL_BIN_NAME} $(CONFIG_ICCM_BASE_ADDRESS)"
+FLASH_SCRIPT = arc_debugger.sh
+DEBUG_SCRIPT = arc_debugger.sh
```
A welcomed helper!

fanghuaqi replied:
Thanks

tbursztyka reviewed (dismissed):
Tiny comment about Copyright. Unless I am mistaken, if you want to update the date keep the very first one and add the new one besides: s/2016/2016-2017

ruuddw commented on arch/arc/core/regular_irq.S:
```diff
@@ -121,7 +121,8 @@ SECTION_FUNC(TEXT, _rirq_exit)
 	 */
 	ldh_s r0, [r2, _thread_offset_to_preempt]
 	mov r3, _NON_PREEMPT_THRESHOLD
-	brhs.d r0, r3, _rirq_no_reschedule
```
What's the reason for this (and next) change? Branch target not in range for the brhs instruction?

vonhust replied:
To support FPU, some instructions are added to save and load fpu related regs. This results the branch target address out of the range of brhs instruction.

ruuddw commented:
I think the EMSK 2.3 platform/bitfile is not fully backwards compatible with EMSK 2.2. Did you consider having 2.2 and 2.3 versions side by side (e.g. as different SoCs or boards) instead of updating 2.2?

dbkinder commented on boards/arc/em_starterkit/doc/board.rst:
```diff
@@ -174,122 +181,55 @@ Parity:    None
+EM Starterkit directly.
 
-For Linux, you will need:
+Boot up the SOC by pressing the "C" button. Be sure the digilent cable
```
to be clear,is this a "C" button on the board?  Also, if the cable needs to be connected before pressing the "C" button, put that step first rather than as an afterthought.

fanghuaqi replied:
Thanks for notification, I will make it more clear.

dbkinder commented on boards/arc/em_starterkit/doc/board.rst:
```diff
@@ -174,122 +181,55 @@ Parity:    None
-
-In any terminal window, invoke the "ocd" script first. It should establish
-contact with the board and output many messages. For example:
+Connect to the debug server at the emsk from a second console:
```
is "emsk" an abbreviation for the EM starter kit?  I'd suggest just spelling it out here since you don't use the term "emsk" anywhere else.

fanghuaqi replied:
Got it. Will fix.

dbkinder commented on boards/arc/em_starterkit/doc/board.rst:
```diff
@@ -174,122 +181,55 @@ Parity:    None
-   target state: halted
-   target state: halted
+   $ cd <my app>
+   $ $ZEPHYR_SDK_INSTALL_DIR/sysroots/x86_64-pokysdk-linux/usr/bin/arc-zephyr-elf/arc-zephyr-elf-gdb outdir/em_starterkit/zephyr.elf
```
you could make this line shorter so it doesn't wrap by breaking the line at a space and put in a backslash

fanghuaqi replied:
Good suggestion. Will fix.

dbkinder commented on boards/arc/em_starterkit/doc/board.rst:
```diff
@@ -174,122 +181,55 @@ Parity:    None
 
-In a second console window, navigate to the directory for your sample application,
-and invoke the debug.sh script:
+If you just want to download the application to the emsk's CCM or DDR and run,
```
found one more "emsk" -- I'd recommend spelling it out "EM starter kit"

fanghuaqi replied:
Yeah, will fix it.

dbkinder commented on boards/arc/em_starterkit/doc/board.rst:
```diff
@@ -174,122 +181,55 @@ Parity:    None
 
-Flashing
-========
+This command still use openocd and gdb to load application elf file to emsk,
```
s/use/uses/

fanghuaqi replied:
Will fix it.

dbkinder reviewed (changes requested):
a few suggestions/questions

fanghuaqi committed:
arc: Fix copyright year mistake in soc.h of arc em7d, em9d, em11d

Change 2016 to 2016-2017. I need to provide the starting date in
copyright header.

Signed-off-by: Huaqi Fang <huaqi.fang@synopsys.com>

fanghuaqi reviewed (commented):
Thanks for reviewing, I just modified the copyright header in soc.h of em7d, em9d, em11d in commit e620d977

fanghuaqi reviewed (commented):
@vonhust  just explained the changes

fanghuaqi committed:
doc: Update documentation for EM Starter Kit

* Change emsk to EM Starter Kit
* Fix some typos
* Add more statement for steps of using EM Starter Kit

Signed-off-by: Huaqi Fang <huaqi.fang@synopsys.com>

fanghuaqi reviewed (commented):
Update EM Starter Kit documentation to make it more clear

vonhust commented:
For Ruud's concern,  in the future, we will support em staterkit 2.2 and em startkit 2.3 together.  User s can make a choice between 2 versions.

fanghuaqi reviewed (commented):
Documentation for EM StarterKit is updated

dbkinder commented on boards/arc/em_starterkit/doc/board.rst:
```diff
@@ -123,6 +132,26 @@ pieces of hardware are required.
+  can download and debug Zephyr application now.
+
+* If you want to know more about how to use this board, you can take a look
+  at the **ARC EM Starter Kit User Guide**.
```
can you provide a link to this user guide

fanghuaqi replied:
OK, I will add it, I thought the document version hosted on https://www.embarc.org/help.html#starterkit was an old one for EM StarterKit, but now it seems it is updated to latest version. So I will add link to it.

dbkinder reviewed (commented):
Looks better, thanks!  You mention a user guide -- can you add a link to it please.

fanghuaqi committed:
doc: boards: Update reference links in em starterkit documentation

+ Fixed some unused links in the em starterkit doc
+ Provided a link for em starterkit user guide

Signed-off-by: Huaqi Fang <huaqi.fang@synopsys.com>

fanghuaqi reviewed (commented):
I just updated links used in this doc

dbkinder reviewed (approved):
+1 for docs
